### DISCUSSION TITLE
APA-Style numbering in Header

### DISCUSSION THREAD

**Discussion by Karstack-overflow** [typst/typst#4547] [Q&A]
I need help creating a template in 7th generation APA style. The APA style specifies that the page numbering is placed in the upper right corner and the title or chapter on the left side. If I now define the page and insert the page number with counter(page).display(), I get the previous page number displayed in the header if I use counter().update() beforehand. For example, it should be 1, but it is 5. In the footer this works strangely enough. The style should be adapted to the number format, e.g. the table of contents should have Roman numerals. I would be very grateful for any help in solving this problem with the page layout. So far my set page is like this:

```
  set page(
    numbering: "1",
    number-align: top,
    margin: (left: 25mm, right: 25mm, top: 30mm, bottom: 30mm),
    header-ascent: 30%, 
    header: context{
      set text(0.9em)
      // Get current page number.
      let page-number = here().page()
      let style = here().page-numbering()
  
      // let page-number =int(counter(page).display())
      if showTitleInHeader != true {
        return []
      }
      // If the current page is the start of a chapter, don't show a header
      let target = heading.where(level: 1)
      if query(target).any(it => it.location().page() == page-number) {
        return [New chapter! page #counter(page).display(style)]
        return []
      }
      // Find the chapter of the section we are currently in.
      let before = query(target.before(here()))
      if before.len() > 0 {
        let current = before.last()

        let chapter-title = current.body
        let chapter-number = counter(heading.where(level: 1)).display()
        // let chapter-number-text = [#current.supplement Chapter #chapter-number]
        let chapter-number-text = [Kapitel #chapter-number: ]
         let (left-text, right-text) = {("", "")}
        if current.numbering != none {
          if draft{
           (left-text, right-text) = {( chapter-number-text + chapter-title , [DRAFT #counter(page).display(style)])}
        }else{
          (left-text, right-text) = {( chapter-number-text + chapter-title , counter(page).display(style))}
        }
          text(weight: "bold", fill-line(left-text, right-text))
          v(-1em)
          line(length: 100%, stroke: 0.5pt)
          
        }
      }
    }
  )
  ```

**Comment by jamesrswift** [comment]
The `Hydra` package is probably what you're looking for https://typst.app/universe/package/hydra/

**Comment by Karstack-overflow** [comment]
thanks. This is a good extension and will make configuration much easier. My problem that after a #counter(page).update(1) the updated page number is not displayed remains. 
![image](https://github.com/user-attachments/assets/96f0d2f0-5789-4e61-b7da-a4e656c7afff)

**Comment by Karstack-overflow** [comment]
when I try to add the page number to the header and have previously updated it. I get the previous one displayed. The correct page number is displayed in the footer.

**Comment by levincode** [comment]
I found this solution to get the page number.
It seems that when you update the page number in the document content you cannot access the value from that position in the header.
That's why you should update the counter from the header.
Since the pages are not locatable, you need to reference an element to work from there.

```typst
#set page(header: context{
  // query a specific element on the page where you want to reset your page counter
  let intro = query(<intro>).first()

  // match with current page
  if intro.location().page() == here().page() {
    counter(page).update(1) // update the counter
  }

  // there you go.
  [
    page: #locate(loc=> counter(page).at(loc).first())
    or: #context counter(page).get().first()
  ]
})
```

Every time you want to update the page number you must do it from the header and in the specific position.
